Шумилин Илья

Jevvry opened the pull request:
@RGleb

Jevvry committed:
Mvp

RGleb commented on TagsCloud/TextProcessing/WordConfig/IWordsConfig.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public interface IWordsConfig
+    {
+        public Font FontName { get; set; }
```
В интерфейсе не обязательно public писать у сигнатуры

RGleb commented on TagsCloud/Tests/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,119 @@
+
+namespace TagsCloud.Tets
+{
+    [TestFixture]
```
Этот атрибут необязательный

RGleb commented on TagsCloud/TextProcessing/TextOperator.cs:
```diff
@@ -0,0 +1,52 @@
+
+        private string ConvertWord(string word) => wordConverters
+                                                    .Aggregate(word, (current, converter) => converter.Convert(current));
+        private ITextReader GetReader(string path)
```
Может этот код вынести в какую-нибудь фабрику?

RGleb commented on TagsCloud/TextProcessing/TextOperator.cs:
```diff
@@ -0,0 +1,52 @@
+            var reader = GetReader(path);
+            var words = reader.ReadWords(path);
+            var filtredWords = words
+                                .Select(ConvertWord)
```
Если возникнет потребность применить не все конвертации, то как это сделать?

RGleb replied:
Если возникнет потребность применить конвертации в определенной последовательности, то как то сделать?

RGleb commented on TagsCloud/TextProcessing/TextOperator.cs:
```diff
@@ -0,0 +1,52 @@
+            var words = reader.ReadWords(path);
+            var filtredWords = words
+                                .Select(ConvertWord)
+                                .Where(word => textFilters.All(filter => filter.CanTake(word)));
```
Если возникнет потребность применить не все фильтры, а только часть из них, то как это сделать?

RGleb commented on TagsCloud/TextProcessing/TextReaders/ITextReader.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+
+namespace TagsCloud.TextProcessing
```
Во многих файлах неймспейсы не соответствуют пути до файла по папкам

RGleb commented on TagsCloud/TextProcessing/TextReaders/ITextReader.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace TagsCloud.TextProcessing
+{
+    public interface ITextReader
```
Скорее IWordsReader

RGleb commented on TagsCloud/TextProcessing/WordInfo.cs:
```diff
@@ -0,0 +1,14 @@
+    public class WordInfo
+    {
+        public string Value { get; private set; }
+        public int Count { get; private set; }
```
Скорее Frequence

RGleb commented on TagsCloud/TextProcessing/TextOperator.cs:
```diff
@@ -0,0 +1,52 @@
+        private readonly ITextFilter[] textFilters;
+        private readonly IWordConverter[] wordConverters;
+
+        public TextOperator(IEnumerable<ITextReader> textReaders, IEnumerable<ITextFilter> textFilters,
```
Контейнер не умеет массивами подставлять аргументы?

RGleb commented on TagsCloud/Layouter/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,70 @@
+        public CircularCloudLayouter()
+        {
+            rectangles = new List<Rectangle>();
+            spiral = new Spiral(new Point(0, 0));
```
Можно же через конструктор принять спираль, да и сделать ее реализацией какого-нибудь интерфейса

RGleb commented:
Хочется еще тесты на полную работу программы

Jevvry committed:
feature:
    - add ConvertsFactory
    - add FiltersFactory
    - add ReadersFactory

Jevvry committed:
feature:
    - add checkBox on ConfigWindow

Jevvry committed:
refactor:
    - add TagsCloudProcessor

Jevvry committed:
refactor:
    - transferring tests to a separate project

Jevvry committed:
refactor:
    - code style
    - naming

Jevvry committed:
refactor:
    - namespace definition
    - delete unnecessary using

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,57 @@
+        public Bitmap CreateCloud()
+        {
+            var imageSize = imageConfig.ImageSize;
+            layouter.SetCenter(new Point(imageSize.Width / 2, imageSize.Height / 2));
```
`.SetCenter` выглядит как костыль

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,57 @@
+            var tags = new List<Tag>();
+
+            var count = 40;
+            foreach (var word in words.Take(30))
```
А если эта логика создания тегов поменяется, то ее здесь придется менять?

RGleb replied:
А если мы захотим разными цветами раскрашивать слова, то где это сделать?

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,57 @@
+            this.imageConfig = imageConfig;
+        }
+
+        public Bitmap CreateCloud()
```
На мой взгляд не очень хорошо, когда IDisposable Bitmap создается внутри метода и возвращается наружу, потому что никто не обещает что он задеспоузится, когда не нужен будет

RGleb replied:
Да и вообще его возвращать не нужно, там где он используется можно по-другому выкрутиться, например мы знаем куда файл с изображением сохранился, поэтому можно вызвать Bitmap.Fromfile

RGleb commented on TagsCloud/TextProcessing/TextOperator.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace TagsCloud.TextProcessing
+{
+    public class TextOperator
```
Почему именно TextOperator? Не совсем понимаю

RGleb commented on TagsCloud/TextProcessing/TextFilters/FiltersFactory.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace TagsCloud.TextProcessing.TextFilters
+{
+    public class FiltersFactory : IFiltersFactory
```
Это не фабрика, это скорее FiltersApplier

RGleb commented on TagsCloud/TextProcessing/TextOperator.cs:
```diff
@@ -0,0 +1,35 @@
+        private readonly IReadersFactory readersFactory;
+        private readonly IFiltersFactory filtersFactory;
+        private readonly IConvertersFactory convertersFactory;
+        private readonly IWordsConfig wordsConfig;
```
На мой взгляд не очень хорошо получилось, что TextOperator зависит от конфига
Когда такая зависимость у каких-то Factory или Applier, то вроде еще норм

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,37 @@
+            services.Scan(scan => scan
+                                    .FromCallingAssembly()
+                                    .AddClasses()
+                                    .AsSelfWithInterfaces());
```
А какое время жизни у объектов тут?

Jevvry replied:
Вроде бы всегда создаются новые состояния, если мы не укажем иное.

RGleb replied:
Давай лучше это явно укажем

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,37 @@
+            Application.Run(BuildContainer().GetService<Form>());
+        }
+
+        public static ServiceProvider BuildContainer()
```
Не настаиваю, но на мой взгляд так красивее смотрится :)
![image](https://user-images.githubusercontent.com/33834646/100638015-55d21480-3355-11eb-8d2d-81d62c76fb72.png)

RGleb commented on TagsCloud/ImageProcessing/Config/IImageConfig.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public interface IImageConfig
+    {
+        public Size ImageSize { get; set; }
```
Зачем пишешь public свойств/методов в интерфесе. У меня это подчеркивается красным здесь и во всех папках включая проект и солюшен
![image](https://user-images.githubusercontent.com/33834646/100638419-d3962000-3355-11eb-9518-5e11dece2c49.png)

RGleb commented on TagsCloud/TextProcessing/TextReaders/ReadersFactory.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace TagsCloud.TextProcessing.TextReaders
+{
+    public class ReadersFactory : IReadersFactory
```
Это не фабрика, это какой-то умный Reader

Jevvry committed:
feature:
    - add LayouterFactory
    - add TagsGenerator
    - add TagsGeneratorFactory

Jevvry committed:
refactor:
    - add base class for Layouter
    - remove Spiral

Jevvry committed:
feature:
    - add support doc, docx extension
    - add support jpg, bmp extension

Jevvry committed:
refactor:
    - code style
    - naming

Jevvry committed:
refactor:
    - add TagsGeneratorFactory register method
    - add ConvertersApplier register method
    - add FiltersApplier register method

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+            Application.Run(container.GetService<Form>());
+        }
+
+        public static ServiceProvider BuildContainer()
```
private

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+                .BuildServiceProvider();
+        }
+
+        private static void InitFactories(ServiceProvider serviceProvider)
```
ConfigureFactories

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+                .Register("По спирали", center => new CircularCloudLayouter(center));
+
+            serviceProvider.GetService<ITagsGeneratorFactory>()
+                .Register("Топ 30", () => new TagsGenerator());
```
serviceProvider.GetService

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+                .Register("Топ 30", () => new TagsGenerator());
+
+            serviceProvider.GetService<IConvertersApplier>()
+                .Register("Перевести в нижний регистр", new LowerCaseConverter());
```
serviceProvider.GetService

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+                .Register("Перевести в нижний регистр", new LowerCaseConverter());
+
+            serviceProvider.GetService<IFiltersApplier>()
+                .Register("Исключить служебные символы", new FunctionWordsFilter());
```
serviceProvider.GetService

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+        public static ServiceProvider BuildContainer()
+        {
+            return new ServiceCollection()
+                .Scan(scan => scan.FromCallingAssembly().AddClasses().AsSelfWithInterfaces())
```
С каким скоупом тут регистрируются объекты?

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,48 @@
+        public void CreateCloud()
+        {
+            var imageSize = imageConfig.ImageSize;
+            var layouter = layouterFactory.Create(new Point(imageSize.Width / 2, imageSize.Height / 2));
```
Точку можно и в фабрике получить из конфига

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,48 @@
+            var tagsGenerator = tagsGeneratorFactory.Create();
+
+            var words = textOperator.ReadFromFile(wordsConfig.Path);
+            var tags = tagsGenerator.CreateTags(words, layouter, wordsConfig.FontName);
```
этот layouter используется только в CreateTags может его создавать лучше tagsGenerator-e?

RGleb replied:
tagsGeneratorFactory.Create().GenerateTags()

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,48 @@
+            this.imageConfig = imageConfig;
+        }
+
+        public void CreateCloud()
```
Кажется, будет понятнее если сюда будут передаваться пути: откуда читать слова, и куда сохранять картинку

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace TagsCloud.TagsCloudProcessing
+{
+    public class TagsCloudProcessor
```
Processor, а единственный метод называется CreateCloud, тогда скорее TagsCloudCreator

RGleb commented on TagsCloud/TextProcessing/TextProcessor.cs:
```diff
@@ -0,0 +1,33 @@
+    public class TextProcessor
+    {
+        private readonly IReadersFactory readersFactory;
+        private readonly IFiltersApplier filtersFactory;
```
applier

RGleb commented on TagsCloud/TextProcessing/TextProcessor.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        private readonly IReadersFactory readersFactory;
+        private readonly IFiltersApplier filtersFactory;
+        private readonly IConvertersApplier convertersFactory;
```
applier

RGleb commented on TagsCloud/TextProcessing/TextProcessor.cs:
```diff
@@ -0,0 +1,33 @@
+        {
+            var reader = readersFactory.CreateReader(path);
+            var words = reader.ReadWords(path);
+            var filtredWords = filtersFactory.ApplyFilters(words);
```
filtered

RGleb commented on TagsCloud/TextProcessing/TextProcessor.cs:
```diff
@@ -0,0 +1,33 @@
+        public IEnumerable<WordInfo> ReadFromFile(string path)
+        {
+            var reader = readersFactory.CreateReader(path);
+            var words = reader.ReadWords(path);
```
readersFactory.CreateReader().ReadWords

RGleb commented on TagsCloud/ImageProcessing/SaverImage/ImageSaver.cs:
```diff
@@ -0,0 +1,31 @@
+        public void SaveImageWithConfig(Bitmap bitmap, IImageConfig imageConfig)
+        {
+            var path = imageConfig.Path;
+            var format = formats.FirstOrDefault(pair => path.EndsWith(pair.Key)).Value;
```
Может этот код тоже в фабрику унести?

RGleb commented on TagsCloud/Layouter/ILayouter.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace TagsCloud.Layouter
+{
+    public interface ILayouter
```
Скорее IRectanglesLayouter

RGleb replied:
И в IRectanglesFactory

RGleb commented on TagsCloud/Layouter/ILayouter.cs:
```diff
@@ -0,0 +1,10 @@
+    public interface ILayouter
+    {
+        Rectangle PutNextRectangle(Size rectangleSize);
+        Size GetLayoutSize();
```
У этого метода нет юсаджей, может удалить его?

RGleb commented on TagsCloud/Layouter/Base/LayouterBase.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            var rectangle = new Rectangle();
+            rectangle.Size = rectangleSize;
+            foreach (var point in GetPoints())
```
На сколько понимаю, точки каждый раз сначала будут считаться, это долго

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,63 @@
+                .AddSingleton<IConvertersApplier, ConvertersApplier>()
+                .AddSingleton<IFiltersApplier, FiltersApplier>()
+
+                .AddScoped<Form, ConfigWindow>()
```
Почему именно Scoped?

RGleb commented on TagsCloud/TagsCloudProcessing/TagsCloudProcessor.cs:
```diff
@@ -0,0 +1,48 @@
+    public class TagsCloudProcessor
+    {
+        private readonly ILayouterFactory layouterFactory;
+        private readonly IWordsConfig wordsConfig;
```
Кажется, конфиги должны отсюда пропасть

Jevvry committed:
refactor:
    - add ImageSaversFactory
    - add Savers
    - add ServiceProviderExtensions
    - naming

Jevvry committed:
refactor:
    - remove unnecessary using

Jevvry committed:
small fix Test Project

RGleb commented on TagsCloud/Layouter/Base/LayouterBase.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace TagsCloud.Layouter.Base
+{
+    public abstract class LayouterBase : IRectanglesLayouter
```
RectanglesLayouterBase

RGleb commented on TagsCloud/Layouter/Base/LayouterBase.cs:
```diff
@@ -0,0 +1,48 @@
+        public LayouterBase(Point center)
+        {
+            this.center = center;
+            rectangles = new List<Rectangle>();
```
Может инициализировать в месте объявление?

RGleb commented on TagsCloud/Layouter/Base/LayouterBase.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public Rectangle PutNextRectangle(Size rectangleSize)
+        {
+            var rectangle = new Rectangle();
```
var rectangle = new Rectangle {Size = rectangleSize}

RGleb commented on TagsCloud/Layouter/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,29 @@
+    public class CircularCloudLayouter : LayouterBase
+    {
+        private readonly Point center;
+        private int parameter;
```
Скорее index

RGleb commented on TagsCloud/ImageProcessing/Config/IImageConfig.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace TagsCloud.ImageProcessing.Config
+{
+    public interface IImageConfig
```
На мой взгляд к конфигу не имеет смысла интерфейс выделять

RGleb replied:
Касается всех конфигов в проекте

RGleb commented on TagsCloud/Extensions/RectangleExtension.cs:
```diff
@@ -0,0 +1,27 @@
+    {
+        public static bool IntersectsWith(this Rectangle rectangle, IEnumerable<Rectangle> rectangles)
+        {
+            return rectangles.Any(otherRect => rectangle.IntersectsWith(otherRect));
```
.Any(rectangle.IntersectsWith)

RGleb commented on TagsCloudTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,124 @@
+        }
+
+        [Test]
+        [Timeout(100000)]
```
Лучше 100 * 1000

RGleb commented on TagsCloudTest/FunctionalTests.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace TagsCloudTest
+{
+    public class FunctionalTests
```
Скорее TagsCloudCreatorTests

RGleb commented on TagsCloudTest/FunctionalTests.cs:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        [Test]
+        public void FunctionalTest()
```
Хочется имя получше

RGleb commented on TagsCloudTest/FunctionalTests.cs:
```diff
@@ -0,0 +1,70 @@
+        {
+            tagsCloudCreator.CreateCloud(textPath, imagePath);
+            using var image = Image.FromFile(imagePath);
+            image.Should().NotBeNull();
```
Хочется еще тест, которые проверит сгенерированную картинку

RGleb commented on TagsCloudTest/Program.cs:
```diff
@@ -0,0 +1,15 @@
+    static class Program
+    {
+        [STAThread]
+        static void Main()
```
Зачем в тестах мэйн? Почему не библиотекой?

RGleb commented on TagsCloudTest/TargetVectorTests.cs:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        [Test]
+        public void PartialDeltaAllDeltaMoveToTraget()
```
*ToTarget

RGleb commented on TagsCloudTest/FunctionalTests.cs:
```diff
@@ -0,0 +1,70 @@
+        public void SetUp()
+        {
+            var path = Assembly.GetExecutingAssembly().Location;
+            var s = new DirectoryInfo(path).Parent.Parent.Parent.FullName;
```
Так не надо делать, есть же ресурсы

RGleb replied:
Касается получения пути к файлам

RGleb commented on TagsCloud/Program.cs:
```diff
@@ -0,0 +1,46 @@
+        private static ServiceProvider BuildContainer()
+        {
+            return new ServiceCollection()
+                .Scan(scan => scan.FromCallingAssembly().AddClasses().AsSelfWithInterfaces().WithTransientLifetime())
```
А может все сервисы зарегистрировать как Singleton, тогда и .AddSingleton не придется писать, или что-то сломается если так сделать?

RGleb commented on TagsCloud/Extensions/ServiceProviderExtensions.cs:
```diff
@@ -0,0 +1,37 @@
+        public static ServiceProvider ConfigureFactories(this ServiceProvider serviceProvider)
+        {
+            serviceProvider.GetService<IImageSaverFactory>()
+                .Register(".png", () => serviceProvider.GetService<PngSaver>())
```
() => serviceProvider.GetService<PngSaver> можно заменить на serviceProvider.GetService<PngSaver>

RGleb replied:
И ниже тоже

RGleb commented on TagsCloud/TextProcessing/WordInfo.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class WordInfo
+    {
+        public string Value { get; private set; }
```
Word

RGleb commented on TagsCloud/ImageProcessing/SaverImage/SaverBase.cs:
```diff
@@ -0,0 +1,16 @@
+namespace TagsCloud.ImageProcessing.SaverImage
+{
+    public abstract class SaverBase : IImageSaver
+    {
```
Может тут сделать также как в CanRead в IWordsReader

RGleb commented on TagsCloud/ImageProcessing/SaverImage/Factory/IImageSaverFactory.cs:
```diff
@@ -0,0 +1,11 @@
+    public interface IImageSaverFactory
+    {
+        IImageSaver Create();
+        IImageSaverFactory Register(string imageEntesion, Func<IImageSaver> creationFunc);
```
imageExtension

Jevvry commented on TagsCloud/Layouter/Base/LayouterBase.cs:
```diff
@@ -0,0 +1,48 @@
+            return rectangle;
+        }
+
+        protected abstract IEnumerable<Point> GetPoints();
```
Алгоритм в цикле перебирает точки, поэтому нужна коллекция

Jevvry replied:
+ в математическом смысле кривой логично, что она возвращает свои точки

RGleb replied:
Здесь получился неочевидный код в плане что когда вызываешь IEnumerable<Point> GetPoints() то ожидаешь что он будет перебирать точки с самого начала, а на самом деле он продолжит перебирать точки после предыдущего своего вызова, поэтому предлагаю сделать Point GetNextPoint(), который уже при каждом вызове будет возвращать следующую точку, и соответственно там где вызывался GetPoints() поменять код, чтобы заработал GetNextPoint

Jevvry committed:
feature:
    - add ServiceFactory

Jevvry committed:
refactor:
    - remove unnecessary factories interfaces
    - remove config interfaces

Jevvry committed:
test:
   - add resources
   - add bitwise comparison for functional test
   - remove unnecessary field in CircularLayouterTests

RGleb commented on TagsCloudTest/TagsCloudCreatorTests.cs:
```diff
@@ -0,0 +1,80 @@
+        }
+
+        [Test]
+        public void TagsCloudFunctionalTest()
```
Падает тест
![image](https://user-images.githubusercontent.com/33834646/101130443-4a1c7180-3625-11eb-8b3d-8cee78081d07.png)

Jevvry replied:
Тест может падать из-за настроек системы. На моей машине проходит [
![img](https://user-images.githubusercontent.com/39431685/101203707-eded3800-368c-11eb-89ad-8ecad444c3d0.jpg)
](url)

RGleb commented on TagsCloudTest/TagsCloudCreatorTests.cs:
```diff
@@ -0,0 +1,80 @@
+        }
+
+        [Test]
+        public void TagsCloudFunctionalTest()
```
Functional абстрактное слово, тут лучше подойдет имя CreateCloud_Test

RGleb commented on TagsCloudTest/TagsCloudCreatorTests.cs:
```diff
@@ -0,0 +1,80 @@
+            var current = File.ReadAllBytes(imagePath);
+            var expected = File.ReadAllBytes(expectedImagePath);
+
+            image.Should().NotBeNull();
```
А нужна ли такая проверка? Какую ситуацию она проверяет?

RGleb commented on TagsCloudTest/TagsCloudCreatorTests.cs:
```diff
@@ -0,0 +1,80 @@
+            var expected = File.ReadAllBytes(expectedImagePath);
+
+            image.Should().NotBeNull();
+            current.Should().BeEquivalentTo(expected);
```
С байтами метод BeEquivalentTo работает как нам нужно: сравнивает соответствующие позиции, а вот с интами и классами уже по-другому: ищет среди всей коллекции совпадающий итем с текущим итемом, на скрине пример, давай лучше Should().Equal использовать, он всегда сравнивает соответствующие позиции
![image](https://user-images.githubusercontent.com/33834646/101131590-5c97aa80-3627-11eb-9bae-0d17b8eb812f.png)

RGleb commented on TagsCloudTest/TagsCloudCreatorTests.cs:
```diff
@@ -0,0 +1,80 @@
+        private TagsCloudCreator tagsCloudCreator;
+        private WordConfig wordsConfig;
+        private ImageConfig imageConfig;
+        private ServiceProvider container = ContainerBuilder.BuildContainer();
```
readonly

Jevvry replied:
Они же инициализируются в SetUp

RGleb replied:
я только про private readonly ServiceProvider

RGleb commented on TagsCloud/TextProcessing/WordsConfig/WordConfig.cs:
```diff
@@ -0,0 +1,16 @@
+        public Font FontName { get; set; }
+        public Color Color { get; set; }
+        public string Path { get; set; }
+        public string[] FilerNames { get; set; }
```
FiltersName

RGleb commented on TagsCloud/TextProcessing/WordsConfig/WordConfig.cs:
```diff
@@ -0,0 +1,16 @@
+        public string Path { get; set; }
+        public string[] FilerNames { get; set; }
+        public string[] ConvertersNames { get; set; }
+        public string LayoutName { get; set; }
```
LayouterName

RGleb commented on TagsCloud/TextProcessing/WordsConfig/WordConfig.cs:
```diff
@@ -0,0 +1,16 @@
+{
+    public class WordConfig
+    {
+        public Font FontName { get; set; }
```
Просто Font

RGleb replied:
Или string FontName

RGleb replied:
Лучше второе

RGleb commented on TagsCloudTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,123 @@
+            return sizes;
+        }
+
+        private void PutRectangles(int rectangleCount, Size[] sizes)
```
Можно проще, кстати тут не обрабатывается ситуация, когда массив sizes будет пустой
![image](https://user-images.githubusercontent.com/33834646/101133661-fdd43000-362a-11eb-98d3-2d6ab5b1dc72.png)

Jevvry replied:
А зачем?

RGleb replied:
зачем проще или зачем обрабатывать?

Jevvry replied:
Зачем обрабатывать?

RGleb replied:
в данном случае можно и не обрабатывать, мы же знаем что sizes не null и не пустая коллекция

RGleb commented on TagsCloudTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,123 @@
+        {
+            PutRectangles(10, GetRandomSizeSet());
+
+            var isIntesect = rectangles
```
isIntersect

RGleb commented on TagsCloudTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,123 @@
+        }
+
+        [Test]
+        public void PutNextRectangleReturnDifferentRectangles()
```
Если записать так, то все случаи проверим? (Size[] в PutRectangles сделал params)
![image](https://user-images.githubusercontent.com/33834646/101134193-ce71f300-362b-11eb-8912-1721e33f511b.png)

Jevvry replied:
Тут важно именно, что они разные и проверка была по эквивалентности с хешсетом

Jevvry replied:
Distinct вроде бы использует сравнение по умолчанию

RGleb replied:
немного не понятно: здесь было примерно так: кладем прямоугольник в лист и в хешсет, и потом проверяем что количество прямоугольников совпадает и что коллекции совпадают. Окей тогда зачем раньше проверялось что количество совпадает, если проверка коллекции проверит и количество тоже. Далее если в хэшсет и в лист кладутся одни и те же прямоугольники, то достаточно проверить что в хэшсете их ровно столько же сколько и в листе. Далее зачем их вручную класть в лист и в хэшсет если PutRectangles уже умеет это делать. Далее в хэшсет не передавался никакой кастомный комперер, поэтому там использовался комперер по умолчанию, т.е. такой же как и в методе Distinct

RGleb replied:
ну ок, можно вызвать метод .ToHashSet

RGleb commented on TagsCloud/TextProcessing/WordInfo.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public class WordInfo
+    {
+        public string Word { get; private set; }
```
private set можно убрать у свойств если они инициализируются в конструкторе

RGleb commented on TagsCloud/TagsCloudProcessing/Tag.cs:
```diff
@@ -0,0 +1,18 @@
+{
+    public class Tag
+    {
+        public string Value { get; set; }
```
Здесь set можно убрать

RGleb commented on TagsCloud/TagsCloudProcessing/TagsGenerators/ITagsGenerator.cs:
```diff
@@ -0,0 +1,10 @@
+﻿using System.Collections.Generic;
+using TagsCloud.TextProcessing;
+
+namespace TagsCloud.TagsCloudProcessing.TegsGenerators
```
Опечатся в неймспейсе "Tegs"

RGleb commented on TagsCloud/TagsCloudProcessing/TagsGenerators/TagsGenerator.cs:
```diff
@@ -0,0 +1,42 @@
+            this.wordsConfig = wordsConfig;
+        }
+
+        public IEnumerable<Tag> CreateTags(IEnumerable<WordInfo> words)
```
Вроде бы так проще и должно работать
![image](https://user-images.githubusercontent.com/33834646/101136767-9e2c5380-362f-11eb-97ac-dcba9c57a423.png)

RGleb commented on TagsCloudTest/ContainerBuilder.cs:
```diff
@@ -0,0 +1,48 @@
+{
+    public static class ContainerBuilder
+    {
+        public static ServiceProvider BuildContainer()
```
Предлагаю тут скопировать код из актуального контейнера

Jevvry committed:
refactor:
    - remove applier interfaces
    - add ServiceFactory for savers
    - add ServiceFactory for readers
    - add ServiceFactory for filters
    - add ServiceFactory for converters

Jevvry committed:
refactor:
    - add CircularCloudLayouterCreator

Jevvry committed:
refactor:
    - project became a library

Jevvry committed:
refactor:
    - code style

Jevvry committed:
refactor:
    - add project for ui

Jevvry committed:
tests:
    - fix container creation
    - code style in CircularCloudLayouterTests

Jevvry committed:
tests:
    - fix CircularCloudLayouterTests
    - fix TagsCloudCreatorTests

Jevvry committed:
remove unnecessary using
